### PR TITLE
[bitnami/zookeeper] Add Zookeeper Helm Chart tests

### PR DESCRIPTION
**Pull request by mdhont**
### Description of the change

Adding the tests for the Zookeeper Helm chart.

### Benefits

When someone makes a change in the Helm Chart, they will be able to check if their change has broken the Helm chart by running these automated tests.

### Checklist

- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: .vib/zookeeper/vib-verify.json [hunk 1]
```diff
@@ -38,6 +38,17 @@
...
+            "resources": {
+              "path": "/.vib/zookeeper/goss"
+            },
+            "remote": {
+              "workload": "sts-zookeeper"
+            }
+          }
+        },
         {
           "action_id": "health-check",
```
Comment (superaleks) [inline .vib/zookeeper/vib-verify.json:53]: Make sure the order of the actions is the same in both `vib-verify` and `vib-publish` pipelines. Right now it´s different. 
1. trivy
2. goss
3. health check

File: .vib/zookeeper/goss/goss.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    contains:
+      - {{ printf "/clientPort=.*%s/" .Env.ZOO_PORT_NUMBER }}
+command:
+  user-id-test:
+    exec: if [ "$(id -u)" -eq 0 ]; then exit 1; fi
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-version-ok:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 version | grep --quiet $(echo $APP_VERSION)
```
Comment (superaleks) [inline .vib/zookeeper/goss/goss.yaml:22]: Is there any way to parametrise the values of the IP or the port?
So that we don´t have to hardcode it.

File: .vib/zookeeper/goss/goss.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+file:
+  /opt/bitnami/zookeeper:
```
Comment (superaleks) [inline .vib/zookeeper/goss/goss.yaml:2]: Not sure if this test adds value, since we are already checking the files inside of this folder, this is then made redundant. WDYT?

File: .vib/zookeeper/goss/goss.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-version-ok:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 version | grep --quiet $(echo $APP_VERSION)
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-create-znode:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /zk_test my_data
```
Comment (superaleks) [inline .vib/zookeeper/goss/goss.yaml:27]: Same comment here as above.

**Review by superaleks — CHANGES_REQUESTED** [review]
Glad to see you contributing @mdhont!

I hope you liked working with Goss. 

The PR is looking very good, I have added some minor comments that you may address as per your preference.

**Comment by mdhont** [comment]
@superaleks, I've made the changes you requested.

File: .vib/zookeeper/goss/goss.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    mode: "0644"
+    contains:
+      - {{ printf "/clientPort=.*%s/" .Env.ZOO_PORT_NUMBER }}
+command:
+  user-id-test:
+    exec: if [ "$(id -u)" -eq 0 ]; then exit 1; fi
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-version-ok:
```
Comment (joancafom) [inline .vib/zookeeper/goss/goss.yaml:18]: You are missing one more test to have a minimum of 5. I recommend you also test if it is possible to create a zNode using the CLI.

File: .vib/zookeeper/goss/goss.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$(echo $ZOO_PORT_NUMBER) version | grep --quiet $(echo $APP_VERSION)
```
[the remark this answers is not in the document]
Comment (joancafom) [reply .vib/zookeeper/goss/goss.yaml:19]: This is not very resilient. The output of `zkCli.sh version` contains a lot of other info about the different jars and components:

```bash
$ /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$ZOO_PORT_NUMBER version | grep $APP_VERSION
2022-07-14 07:29:37,020 [myid:] - INFO  [main:o.a.z.Environment@98] - Client environment:zookeeper.version=3.8.0-5a02a05eddb59aee6ac762f7ea82e92a68eb9c0f, built on 2022-02-25 08:49 UTC
2022-07-14 07:29:37,023 [myid:] - INFO  [main:o.a.z.Environment@98] - Client environment:java.class.path=/opt/bitnami/zookeeper/bin/../zookeeper-server/target/classes:/opt/bitnami/zookeeper/bin/../build/classes:/opt/bitnami/zookeeper/bin/../zookeeper-server/target/lib/*.jar:/opt/bitnami/zookeeper/bin/../build/lib/*.jar:/opt/bitnami/zookeeper/bin/../lib/zookeeper-prometheus-metrics-3.8.0.jar:/opt/bitnami/zookeeper/bin/../lib/zookeeper-jute-3.8.0.jar:/opt/bitnami/zookeeper/bin/../lib/zookeeper-3.8.0.jar:/opt/bitnami/zookeeper/bin/../lib/snappy-java-1.1.7.7.jar:/opt/bitnami/zookeeper/bin/../lib/slf4j-api-1.7.30.jar:/opt/bitnami/zookeeper/bin/../lib/simpleclient_servlet-0.9.0.jar:/opt/bitnami/zookeeper/bin/../lib/simpleclient_hotspot-0.9.0.jar:/opt/bitnami/zookeeper/bin/../lib/simpleclient_common-0.9.0.jar:/opt/bitnami/zookeeper/bin/../lib/simpleclient-0.9.0.jar:/opt/bitnami/zookeeper/bin/../lib/netty-transport-native-unix-common-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-transport-native-epoll-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-transport-classes-epoll-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-transport-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-tcnative-classes-2.0.48.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-tcnative-2.0.48.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-resolver-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-handler-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-common-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-codec-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/netty-buffer-4.1.73.Final.jar:/opt/bitnami/zookeeper/bin/../lib/metrics-core-4.1.12.1.jar:/opt/bitnami/zookeeper/bin/../lib/logback-core-1.2.10.jar:/opt/bitnami/zookeeper/bin/../lib/logback-classic-1.2.10.jar:/opt/bitnami/zookeeper/bin/../lib/jline-2.14.6.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-util-ajax-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-util-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-servlet-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-server-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-security-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-io-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/jetty-http-9.4.43.v20210629.jar:/opt/bitnami/zookeeper/bin/../lib/javax.servlet-api-3.1.0.jar:/opt/bitnami/zookeeper/bin/../lib/jackson-databind-2.13.1.jar:/opt/bitnami/zookeeper/bin/../lib/jackson-core-2.13.1.jar:/opt/bitnami/zookeeper/bin/../lib/jackson-annotations-2.13.1.jar:/opt/bitnami/zookeeper/bin/../lib/commons-io-2.11.0.jar:/opt/bitnami/zookeeper/bin/../lib/commons-cli-1.4.jar:/opt/bitnami/zookeeper/bin/../lib/audience-annotations-0.12.0.jar:/opt/bitnami/zookeeper/bin/../zookeeper-*.jar:/opt/bitnami/zookeeper/bin/../zookeeper-server/src/main/resources/lib/*.jar:/opt/bitnami/zookeeper/bin/../conf:
ZooKeeper CLI version: 3.8.0-5a02a05eddb59aee6ac762f7ea82e92a68eb9c0f, built on 2022-02-25 08:49 UTC
```

It may be the case that $APP_VERSION matches with one of them instead! Either restrict the output verbosity if that is possible or use a more restrictive regex.

File: .vib/zookeeper/goss/goss.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-version-ok:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$ZOO_PORT_NUMBER version | grep "ZooKeeper CLI version" | cut -d- -f1 | grep --quiet $APP_VERSION
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-create-znode:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /zk_test my_data
```
[the remark this answers is not in the document]
Comment (joancafom) [reply .vib/zookeeper/goss/goss.yaml:24]: We have to ensure that tests can be executed multiple times. In order to do that, we randomize the data we create. 

You will have to do the same with `zk_test` or else you will find issues on a second run. Take a look at how we did it with Kafka:

https://github.com/bitnami/charts/blob/be2668ac3b3fb7d15211ba6b763f052d74096b02/.vib/kafka/goss/goss.yaml#L3

File: .vib/zookeeper/goss/goss.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    exit-status: 0
+    stdout: []
```
Comment (joancafom) [inline .vib/zookeeper/goss/goss.yaml:26]: It would also be interesting checking that stdout contains a confirmation of the creation, such as:

```
Created /zk_test 
```

This is something that we usually do, take a look at Kafka for instance:

https://github.com/bitnami/charts/blob/be2668ac3b3fb7d15211ba6b763f052d74096b02/.vib/kafka/goss/goss.yaml#L6-L7

**Comment by mdhont** [comment]
3 succesful test runs:

https://github.com/mdhont/charts/pull/5/commits/883f91309d2af9cedd09bcde18d752181f0c1b57
https://github.com/mdhont/charts/pull/5/commits/61b46dbdb8813e4522c6f8d165f7a84ceb7fd91b
https://github.com/mdhont/charts/pull/5/commits/127d423bd0e065011dddb59357312b8976d1070e

**Review by FraPazGal — DISMISSED** [review]
LGTM! Thanks for the PR @mdhont!

File: .vib/zookeeper/goss/goss.yaml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    contains:
+      - {{ printf "/clientPort=.*%s/" .Env.ZOO_PORT_NUMBER }}
+command:
+  user-id-test:
+    exec: if [ "$(id -u)" -eq 0 ]; then exit 1; fi
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-version-ok:
+    exec: "/opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$ZOO_PORT_NUMBER version | grep \"ZooKeeper CLI version: $APP_VERSION\""
```
Comment (joancafom) [inline .vib/zookeeper/goss/goss.yaml:19]: Is there any reason to enclose this with quotes?

```suggestion
    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$ZOO_PORT_NUMBER version | grep "ZooKeeper CLI version: $APP_VERSION"
```

File: .vib/zookeeper/goss/goss.yaml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    exit-status: 0
+    stdout: []
+    stderr: []
+  zookeeper-create-znode:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$ZOO_PORT_NUMBER create /my_key_{{randAlpha 5}} my_data
+    exit-status: 0
+    stdout: []
+    stderr:
+      - "/Created /my_key_[A-Za-z]/"
```
Comment (joancafom) [inline .vib/zookeeper/goss/goss.yaml:28]: It is a RegEx, you don't need to use quotes

```suggestion
      - /Created \/my_key_[A-Za-z]+/
```

**Review by joancafom — COMMENTED** [review]
Good job! Just some minor comments about formatting and we're good to go

File: .vib/zookeeper/goss/goss.yaml (hunk 9 above)
Comment (mdhont) [reply .vib/zookeeper/goss/goss.yaml:19]: I got the "Nested mappings are not allowed in compact mappings" linter error if I don't use the quotes.

Comment (joancafom) [reply .vib/zookeeper/goss/goss.yaml:19]: Then it is okay to use them haha, thanks for checking !

**Review by joancafom — APPROVED** [review]
LGTM!

**Review by FraPazGal — APPROVED** [review]

**Review by superaleks — APPROVED** [review]
LGTM!

**Review by agomezmoron — APPROVED** [review]
LGTM

**Review by FraPazGal — CHANGES_REQUESTED** [review]
The current approvals are no longer valid. We need to revisit these tests following the new approach.

**Comment by mdhont** [comment]
Successful test run:
https://github.com/mdhont/charts/pull/8

File: .vib/zookeeper/goss/goss.yaml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+command:
```
Comment (FraPazGal) [inline .vib/zookeeper/goss/goss.yaml:1]: I'm missing a folder check for the `data` mounted volume. Additionally, is there any conf file where we could check any of the set params on `values.yaml`?

File: .vib/zookeeper/goss/goss.yaml [hunk 12]
```diff
@@ -0,0 +1,4 @@
...
+  zookeeper-create-znode:
+    exec: /opt/bitnami/zookeeper/bin/zkCli.sh -server 127.0.0.1:$ZOO_PORT_NUMBER create /my_key_{{randAlpha 5}} my_data
```
Comment (FraPazGal) [inline .vib/zookeeper/goss/goss.yaml:3]: Could you please modify this test so `ZOO_PORT_NUMBER` is replaced by a vars obtained from `containerPorts.xxxxx` and hard coded in the runtime parameters? I would also try using the K8s service name here instead of `127.0.0.1` to also test that chart's resource as recommended for Spark.

About using the absolute path for the script, if I'm not mistaken it should be in the container's PATH. If we run the tests directly against `zkCli.sh -server...` we would be also checking the PATH is correctly configured.

**Review by agomezmoron — DISMISSED** [review]
LGTM

**Review by FraPazGal — COMMENTED** [review]
Hi @mdhont, I added a couple of comments to the PR.

Another concern is the used `runtime_parameters` for the VIB pipelines. It looks like we are hardcoding values related to auth and TLS that I think we are not using (TLS is disabled by default) nor testing. Same thing with the `logLevel` param; not sure why we are modifying the default value, but we should check it against the asset config files.

**Comment by mdhont** [comment]
Successful test run:

https://github.com/mdhont/charts/runs/7687339839

These latest changes include checking for value that is set in values.yaml and using Goss variables.

File: .vib/zookeeper/goss/goss.yaml (hunk 11 above)
Comment (mdhont) [reply .vib/zookeeper/goss/goss.yaml:12]: I've added a check for the data directory but it's empty. The cfg file is also mounted and the test is checking the contents of that as well. (for a value that is set in the values.yaml)

File: .vib/zookeeper/goss/goss.yaml [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+  /opt/bitnami/zookeeper/conf/zoo.cfg:
+    exists: true
+    filetype: file
+    mode: "0644"
+    contains:
+      - {{ .Vars.containerPorts.client }}
+      - preAllocSize={{ .Vars.preAllocSize }}
+command:
+  zookeeper-create-znode:
+    exec: zkCli.sh -server zookeeper-0:{{ .Vars.containerPorts.client }} create /my_key_{{randAlpha 5}} my_data
```
Comment (mdhont) [inline .vib/zookeeper/goss/goss.yaml:14]: @FraPazGal  I've used a relative path and Goss vars

File: .vib/zookeeper/goss/goss.yaml (hunk 12 above)
Comment (mdhont) [reply .vib/zookeeper/goss/goss.yaml]: This has been improved

**Review by FraPazGal — APPROVED** [review]
LGTM

**Review by agomezmoron — APPROVED** [review]
LGTM
